tailcall-macros/src/resolver.rs

```diff
@@ -0,0 +1,92 @@
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, Data, DeriveInput, Fields};
+
+pub fn expand_resolver_derive(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    let name = &input.ident;
+
+    let variants = if let Data::Enum(data_enum) = &input.data {
+        data_enum
+            .variants
+            .iter()
+            .map(|variant| {
+                let variant_name = &variant.ident;
+                let ty = match &variant.fields {
+                    Fields::Unnamed(fields) if fields.unnamed.len() == 1 => &fields.unnamed[0].ty,
+                    _ => panic!("Resolver variants must have exactly one unnamed field"),
+                };
+
+                (variant_name, ty)
+            })
+            .collect::<Vec<_>>()
+    } else {
+        panic!("Resolver can only be derived for enums");
+    };
+
+    let variant_parsers = variants.iter().map(|(variant_name, ty)| {
+        quote! {
+            let temp_result = <#ty>::from_directives(directives.iter());
+            let result_option = temp_result.to_result();
```

Comment:
why convert to result and then back to Valid instead of using Valid in the first place?